// machinelearn/preprocessing/Imputer.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var _ = __importStar(require("lodash"));
var ops_1 = require("../ops");
var MathExtra_1 = __importDefault(require("../utils/MathExtra"));
/**
 * Imputation transformer for completing missing values.
 *
 * @example
 * import { Imputer } from 'preprocessing/Imputer';
 *
 * const testX = [[1, 2], [null, 3], [7, 6]];
 * const imp = new Imputer({ missingValues: null, axis: 0 });
 * imp.fit(testX);
 * const impResult = imp.fit_transform([[null, 2], [6, null], [7, 6]]);
 * // [ [ 4, 2 ], [ 6, 3.6666666666666665 ], [ 7, 6 ] ]
 */
var Imputer = /** @class */ (function () {
    /**
     *
     * @param {any} missingValues - Target missing value to impute
     * @param {any} strategy - Missing value replacement strategy
     * @param {any} axis - Direction to impute
     * @param {any} copy - To clone the input value
     */
    function Imputer(_a) {
        var _b = _a.missingValues, missingValues = _b === void 0 ? null : _b, _c = _a.strategy, strategy = _c === void 0 ? 'mean' : _c, _d = _a.axis, axis = _d === void 0 ? 0 : _d, 
        // verbose = 0,
        _e = _a.copy, 
        // verbose = 0,
        copy = _e === void 0 ? false : _e;
        var _this = this;
        /**
         * Calculate array of numbers as array of mean values
         * Examples:
         * [ [ 1, 2 ], [ null, 3 ], [ 123, 3 ] ]
         * => [ 1.5, 3, 63 ]
         *
         * [ [ 1, 123 ], [ 2, 3, 3 ] ]
         * => [ 62, 2.6666666666666665 ]
         *
         * @param matrix
         * @param {string[]} steps
         */
        this.calcArrayMean = function (matrix, steps) {
            // TODO: Fix any return type
            // TODO: Fix matrix type any
            return _.reduce(steps, function (result, step) {
                switch (step) {
                    case 'flatten':
                        return _.map(result, _.flatten);
                    case 'filter':
                        return _.map(result, 
                        // Expecting any type of matrics array
                        // TODO: implement a correct type
                        function (arr) {
                            return _.filter(arr, function (z) { return z !== _this.missingValues; });
                        });
                    case 'mean':
                        return _.map(result, _.mean);
                    default:
                        return result;
                }
            }, matrix);
        };
        this.missingValues = missingValues;
        this.strategy = strategy;
        this.axis = axis;
        // this.verbose = verbose;
        this.copy = copy;
        this.means = [];
    }
    /**
     * Fit the imputer on X.
     * @param {any[]} X - Input data in array or sparse matrix format
     */
    Imputer.prototype.fit = function (X) {
        if (X === void 0) { X = null; }
        ops_1.validateMatrix2D(X);
        var _X = this.copy ? _.clone(X) : X;
        var rowLen = MathExtra_1.default.size(_X, 0);
        var colLen = MathExtra_1.default.size(_X, 1);
        var rowRange = MathExtra_1.default.range(0, rowLen);
        var colRange = MathExtra_1.default.range(0, colLen);
        if (this.strategy === 'mean') {
            if (this.axis === 0) {
                var colNumbers = _.map(colRange, function (col) {
                    return MathExtra_1.default.subset(_X, rowRange, [col]);
                });
                this.means = this.calcArrayMean(colNumbers, [
                    'flatten',
                    'filter',
                    'mean'
                ]);
            }
            else if (this.axis === 1) {
                var rowNumbers = _.map(rowRange, function (row) { return _.get(_X, "[" + row + "]"); });
                this.means = this.calcArrayMean(rowNumbers, ['filter', 'mean']);
            }
        }
        else {
            throw new Error("Unsupported strategy " + this.strategy + " was given");
        }
    };
    /**
     * Fit to data, then transform it.
     * @param {any[]} X - Input data in array or sparse matrix format
     * @returns {any[]}
     */
    Imputer.prototype.fit_transform = function (X) {
        if (X === void 0) { X = null; }
        ops_1.validateMatrix2D(X);
        var _X = _.clone(X);
        if (this.strategy === 'mean' && this.axis === 0) {
            // Mean column direction transform
            for (var row = 0; row < _.size(_X); row++) {
                for (var col = 0; col < _.size(_X[row]); col++) {
                    var value = _X[row][col];
                    _X[row][col] = value === this.missingValues ? this.means[row] : value;
                }
            }
        }
        else if (this.strategy === 'mean' && this.axis === 1) {
            // Mean row direction transform
            for (var row = 0; row < _.size(_X); row++) {
                for (var col = 0; col < _.size(_X[row]); col++) {
                    var value = _X[row][col];
                    _X[row][col] = value === this.missingValues ? this.means[col] : value;
                }
            }
        }
        else {
            throw new Error("Unknown transformation with strategy " + this.strategy + " and axis " + this.axis);
        }
        return _X;
    };
    return Imputer;
}());
exports.Imputer = Imputer;
