// machinelearn/feature_extraction/text.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var _ = __importStar(require("lodash"));
var sw = __importStar(require("stopword"));
var ops_1 = require("../ops");
var nlp_1 = require("../utils/nlp");
var stop_words_1 = require("./stop_words");
/**
 * The CountVectorizer provides a simple way to both tokenize a collection
 * of text documents and build a vocabulary of known words, but also
 * to encode new documents using that vocabulary.
 *
 * @example
 * import { CountVectorizer } from 'machinelearn/feature_extraction';
 *
 * const corpus = ['deep learning ian good fellow learning jason shin shin', 'yoshua bengio'];
 * const vocabCounts = cv.fit_transform(corpus);
 * console.log(vocabCounts); // [ [ 0, 1, 1, 1, 1, 1, 2, 2, 0 ], [ 1, 0, 0, 0, 0, 0, 0, 0, 1 ] ]
 * console.log(cv.vocabulary); // { bengio: 0, deep: 1, fellow: 2, good: 3, ian: 4, jason: 5, learning: 6, shin: 7, yoshua: 8 }
 * console.log(cv.getFeatureNames()); // [ 'bengio', 'deep', 'fellow', 'good', 'ian', 'jason', 'learning', 'shin', 'yoshua' ]
 *
 * const newVocabCounts = cv.transform(['ian good fellow jason duuog']);
 * console.log(newVocabCounts); // [ [ 0, 0, 1, 1, 1, 1, 0, 0, 0 ] ]
 */
var CountVectorizer = /** @class */ (function () {
    function CountVectorizer() {
        this.vocabulary = {};
    }
    /**
     * Learn a vocabulary dictionary of all tokens in the raw documents.
     * @param {string[]} doc - An array of strings
     * @returns {CountVectorizer}
     */
    CountVectorizer.prototype.fit = function (doc) {
        if (doc === void 0) { doc = null; }
        ops_1.validateMatrix1D(doc);
        this.fit_transform(doc);
        return this;
    };
    /**
     * fit transform applies
     * @param {string[]} doc - An array of strings
     * @returns {number[][]}
     */
    CountVectorizer.prototype.fit_transform = function (doc) {
        if (doc === void 0) { doc = null; }
        ops_1.validateMatrix1D(doc);
        var _a = this.buildVocabulary(doc), internalVocabulary = _a.internalVocabulary, pubVocabulary = _a.pubVocabulary;
        this.vocabulary = pubVocabulary;
        this.internalVocabulary = internalVocabulary;
        return this.countVocab(doc);
    };
    /**
     * Transform documents to document-term matrix.
     * Extract token counts out of raw text documents using the vocabulary
     * fitted with fit or the one provided to the constructor.
     * @param {string[]} doc - An array of strings
     * @returns {number[][]}
     */
    CountVectorizer.prototype.transform = function (doc) {
        if (doc === void 0) { doc = null; }
        ops_1.validateMatrix1D(doc);
        return this.countVocab(doc);
    };
    /**
     * Array mapping from feature integer indices to feature name
     * @returns {Object}
     */
    CountVectorizer.prototype.getFeatureNames = function () {
        if (!this.internalVocabulary) {
            throw new Error('You must fit a document first before you can retrieve the feature names!');
        }
        return this.internalVocabulary;
    };
    /**
     * Build a tokenizer/vectorizer
     * @returns {(x: string) => string[]}
     */
    CountVectorizer.prototype.buildAnalyzer = function () {
        var _this = this;
        return function (x) { return _this.preprocess(x, { removeSW: true }); };
    };
    /**
     * Calculates list of vocabularies in the entire document and come up with
     * vocab: index pairs
     * @param doc
     */
    CountVectorizer.prototype.buildVocabulary = function (doc) {
        var analyze = this.buildAnalyzer();
        var processedDoc = _.flowRight(function (d) { return _.uniq(d); }, function (d) { return _.sortBy(d, function (z) { return z; }); }, function (d) { return _.flatten(d); }, function (d) { return _.map(d, function (text) { return analyze(text); }); })(doc);
        var pubVocabulary = _.reduce(processedDoc, function (sum, val, index) {
            return _.set(sum, val, index);
        }, {});
        return {
            internalVocabulary: processedDoc,
            pubVocabulary: pubVocabulary
        };
    };
    /**
     * @ignore
     * Counting number of vocab occurences in the current token of a sentence
     * ['yoshua', 'bengio', 'deep', 'learning'] = vocabulary
     * ['yohua', 'bengio'] => tokens
     * results in
     * [1, 1, 0, 0]
     * @param doc
     */
    CountVectorizer.prototype.countVocab = function (doc) {
        var _this = this;
        var analyze = this.buildAnalyzer();
        // 1. Reducing the doc
        return _.reduce(doc, function (sum, text) {
            var tokens = analyze(text);
            // 2. Looping each vocab for counting
            var sentenceCounted = _.reduce(_this.internalVocabulary, function (sentenceCounts, vocab) {
                // 3. Getting number of occurences of vocab in each tokens (tokens of a sentence)
                var vocabCount = _.reduce(tokens, function (tokenCounts, t) {
                    if (_.isEqual(t, vocab)) {
                        return tokenCounts + 1;
                    }
                    else {
                        return tokenCounts;
                    }
                }, 0);
                return _.concat(sentenceCounts, [vocabCount]);
            }, []);
            return _.concat(sum, [sentenceCounted]);
        }, []);
    };
    /**
     * @ignore
     * preprocess a line of text by applying
     * 1) tokenization
     * 2) removing stopwords
     * @param text
     * @param {any} removeSW
     * @returns {any}
     */
    CountVectorizer.prototype.preprocess = function (text, _a) {
        var _b = _a.removeSW, removeSW = _b === void 0 ? false : _b;
        var tokenizer = new nlp_1.WordTokenizer();
        return _.flowRight(function (x) { return tokenizer.tokenize(x); }, function (x) { return x.join(' '); }, 
        // TODO: Somehow it's removing too many words??!!
        function (x) {
            return removeSW ? sw.removeStopwords(x, stop_words_1.ENGLISH_STOP_WORDS) : x;
        }, function (x) { return x.split(' '); })(text);
    };
    return CountVectorizer;
}());
exports.CountVectorizer = CountVectorizer;
