// machinelearn/module/lib/naive_bayes/gaussian.js
import * as tf from '@tensorflow/tfjs';
import { zip } from 'lodash';
import { reshape, validateFitInputs, validateMatrix2D } from '../ops';
const SQRT_2PI = Math.sqrt(Math.PI * 2);
/**
 * The Naive is an intuitive method that uses probabilistic of each attribute
 * being in each class to make a prediction. It uses Gaussian function to estimate
 * probability of a given class.
 *
 * @example
 * import { GaussianNB } from 'machinelearn/naive_bayes';
 *
 * const nb = new GaussianNB();
 * const X = [[1, 20], [2, 21], [3, 22], [4, 22]];
 * const y = [1, 0, 1, 0];
 * nb.fit({ X, y });
 * nb.predict({ X: [[1, 20]] }); // returns [ 1 ]
 *
 */
export class GaussianNB {
    /**
     * @param X - array-like or sparse matrix of shape = [n_samples, n_features]
     * @param y - array-like, shape = [n_samples] or [n_samples, n_outputs]
     */
    fit(X = null, y = null) {
        validateFitInputs(X, y);
        const { classCategories, mean, variance } = this.fitModel(X, y);
        this.classCategories = classCategories;
        this.mean = mean;
        this.variance = variance;
    }
    /**
     * @param X - array-like, shape = [n_samples, n_features]
     */
    predict(X = null) {
        validateMatrix2D(X);
        return X.map((x) => this.singlePredict(x));
    }
    /**
     * Restore the model from saved states
     * @param modelState
     */
    fromJSON({ classCategories = null, mean = null, variance = null }) {
        this.classCategories = classCategories;
        this.mean = tf.tensor2d(mean);
        this.variance = tf.tensor2d(variance);
    }
    /**
     * Save the model's states
     */
    toJSON() {
        return {
            classCategories: this.classCategories,
            mean: reshape([...this.mean.dataSync()], this.mean.shape),
            variance: reshape([...this.variance.dataSync()], this.variance.shape)
        };
    }
    /**
     * Make a single prediction
     *
     * @param  {ReadonlyArray<number>} X- values to predict in Matrix format
     * @returns T
     */
    singlePredict(X) {
        const matrixX = tf.tensor1d(X, 'float32');
        const numFeatures = matrixX.shape[0];
        // Comparing input and summary shapes
        const summaryLength = this.mean.shape[1];
        if (numFeatures !== summaryLength) {
            throw new Error(`Prediction input ${matrixX.shape[0]} length must be equal or less than summary length ${summaryLength}`);
        }
        const meanValPow = matrixX
            .sub(this.mean)
            .pow(tf.scalar(2))
            .mul(tf.scalar(-1));
        const exponent = meanValPow
            .div(this.variance.mul(tf.scalar(2)))
            .exp();
        const innerDiv = tf.scalar(SQRT_2PI).mul(this.variance.sqrt());
        const probabilityArray = tf
            .scalar(1)
            .div(innerDiv)
            .mul(exponent);
        const selectionIndex = probabilityArray
            .prod(1)
            .argMax()
            .dataSync()[0];
        return this.classCategories[selectionIndex];
    }
    /**
     * Summarise the dataset per class using "probability density function"
     *
     * @param  {Type2DMatrix<number>} X
     * @param  {ReadonlyArray<T>} y
     * @returns InterfaceFitModel
     */
    fitModel(X, y) {
        const classCategories = [...new Set(y)].sort();
        // Separates X by classes specified by y argument
        const separatedByCategory = zip(X, y).reduce((groups, [row, category]) => {
            groups[category.toString()] = groups[category.toString()] || [];
            groups[category.toString()].push(row);
            return groups;
        }, {});
        const momentStack = classCategories.map((category) => {
            const classFeatures = tf.tensor2d(separatedByCategory[category.toString()], null, 'float32');
            return tf.moments(classFeatures, [0]);
        });
        // For every class we have a mean and variance for each feature
        const mean = tf.stack(momentStack.map(m => m.mean));
        const variance = tf.stack(momentStack.map(m => m.variance));
        // TODO check for NaN or 0 variance
        // setTimeout(() => {
        //   if ([...variance.dataSync()].some(i => i === 0)) {
        //     console.error('No variance on one of the features. Errors may result.');
        //   }
        // }, 100);
        return {
            classCategories,
            mean,
            variance
        };
    }
}
