// machinelearn/module/lib/naive_bayes/multinomial.js
import * as tf from '@tensorflow/tfjs';
import { countBy, isEmpty, zip } from 'lodash';
import { reshape, validateFitInputs, validateMatrix2D } from '../ops';
/**
 * Multinomial naive bayes machine learning algorithm
 *
 * The Naive is an intuitive method that uses probabilistic of each attribute
 * being in each class to make a prediction. It uses multinomial function to estimate
 * probability of a given class.
 *
 * @example
 * import { MultinomialNB } from 'machinelearn/naive_bayes';
 *
 * const nb = new MultinomialNB();
 * const X = [[1, 20], [2, 21], [3, 22], [4, 22]];
 * const y = [1, 0, 1, 0];
 * nb.fit({ X, y });
 * nb.predict({ X: [[1, 20]] }); // returns [ 1 ]
 *
 */
export class MultinomialNB {
    constructor() {
        this.alpha = 1;
    }
    // constructor(private readonly alpha: number = 1) {}
    /**
     * Fit date to build Gaussian Distribution summary
     *
     * @param  {Type2DMatrix<number>} X - training values
     * @param  {ReadonlyArray<T>} y - target values
     * @returns void
     */
    fit(X = null, y = null) {
        validateFitInputs(X, y);
        const { classCategories, multinomialDist, priorProbability } = this.fitModel(X, y);
        this.classCategories = classCategories;
        this.multinomialDist = multinomialDist;
        this.priorProbability = priorProbability;
    }
    /**
     * Predict multiple rows
     *
     * @param  {Type2DMatrix<number>} X - values to predict in Matrix format
     * @returns T
     */
    predict(X = null) {
        validateMatrix2D(X);
        if (isEmpty(this.classCategories) ||
            isEmpty(this.multinomialDist) ||
            isEmpty(this.priorProbability)) {
            throw new TypeError('You should fit the model first before running the predict!');
        }
        return X.map(x => this.singlePredict(x));
    }
    /**
     * Returns a model checkpoint
     *
     * @returns InterfaceFitModelAsArray
     */
    toJSON() {
        return {
            classCategories: Array.from(this.classCategories),
            priorProbability: Array.from(this.priorProbability.dataSync()),
            multinomialDist: reshape(Array.from(this.multinomialDist.dataSync()), this.multinomialDist.shape)
        };
    }
    /**
     * Restore the model from states
     * @param multinomialDist - Multinomial distribution values over classes
     * @param priorProbability - Learned prior class probabilities
     * @param classCategories - List of unique class categories
     */
    fromJSON({ multinomialDist = null, priorProbability = null, classCategories = null } = {
        multinomialDist: null,
        priorProbability: null,
        classCategories: null
    }) {
        this.classCategories = classCategories;
        this.priorProbability = tf.tensor1d(priorProbability);
        this.multinomialDist = tf.tensor2d(multinomialDist);
    }
    /**
     * Make a prediction
     *
     * @param  {ReadonlyArray<number>} predictRow
     * @returns T
     */
    singlePredict(predictRow) {
        const matrixX = tf.tensor1d(predictRow, 'float32');
        const numFeatures = matrixX.shape[0];
        const summaryLength = this.multinomialDist.shape[1];
        // Comparing input and summary shapes
        if (numFeatures !== summaryLength) {
            throw new Error(`Prediction input ${matrixX.shape[0]} length must be equal or less than summary length ${summaryLength}`);
        }
        // log is important to use different multinomial formula instead of the factorial formula
        // The multinomial naive Bayes classifier becomes a linear
        // classifier when expressed in log-space
        // const priorProbability = Math.log(1 / classCount);
        const fitProbabilites = this.multinomialDist
            .clone()
            .mul(matrixX);
        // sum(1) is summing columns
        const allProbabilities = fitProbabilites
            .sum(1)
            .add(this.priorProbability);
        const selectionIndex = allProbabilities.argMax().dataSync()[0];
        allProbabilities.dispose();
        return this.classCategories[selectionIndex];
    }
    /**
     * Summarise the dataset per class
     *
     * @param  {Type2DMatrix<number>} X - input distribution
     * @param  {ReadonlyArray<T>} y - classes to train
     */
    fitModel(X, y) {
        const classCounts = countBy(y);
        const classCategories = Array.from(new Set(y));
        const numFeatures = X[0].length;
        const separatedByCategory = zip(X, y).reduce((groups, [row, category]) => {
            if (!(category.toString() in groups)) {
                groups[category.toString()] = [];
            }
            groups[category.toString()].push(tf.tensor1d(row, 'float32'));
            return groups;
        }, {});
        const frequencySumByClass = tf.stack(classCategories.map((category) => tf.addN(separatedByCategory[category.toString()])));
        const productReducedRow = Array.from(frequencySumByClass.sum(1).dataSync());
        // A class's prior may be calculated by assuming equiprobable classes
        // (i.e., priors = (number of samples in the class) / (total number of samples))
        const priorProbability = tf
            .tensor1d(classCategories.map(c => classCounts[c.toString()] / y.length), 'float32')
            .log();
        // log transform to use linear multinomial forumla
        const multinomialDist = frequencySumByClass
            .add(tf.scalar(this.alpha))
            .div(tf
            .tensor2d(productReducedRow, [frequencySumByClass.shape[0], 1], 'float32')
            .add(tf.scalar(numFeatures * this.alpha)))
            .log();
        return {
            classCategories,
            multinomialDist,
            priorProbability
        };
    }
}
