// machinelearn/decomposition/pca.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var tf = __importStar(require("@tensorflow/tfjs"));
var numeric = __importStar(require("numeric"));
var ops_1 = require("../ops");
/**
 * Principal component analysis (PCA)
 *
 * Linear dimensionality reduction using Singular Value Decomposition of
 * the data to project it to a lower dimensional space.
 *
 * - It uses the LAPACK implementation of the full SVD
 * - or randomized a randomised truncated SVD by the method of
 * Halko et al. 2009, depending on the shape
 * of the input data and the number of components to extract. (Will be implemented)
 *
 * @example
 * import { PCA } from 'machinelearn/decomposition';
 *
 * const pca = new PCA();
 * const X = [[1, 2], [3, 4], [5, 6]];
 * pca.fit(X);
 * console.log(pca.components); // result: [ [ 0.7071067811865476, 0.7071067811865474 ], [ 0.7071067811865474, -0.7071067811865476 ] ]
 * console.log(pca.explained_variance); // result: [ [ -0.3535533905932736, 0 ], [ 0, 0.5 ], [ 0.35355339059327373, 0 ] ]
 */
var PCA = /** @class */ (function () {
    function PCA() {
    }
    /**
     * Fit the model with X.
     * At the moment it does not take n_components into consideration
     * so it will only calculate Singular value decomposition
     * @param {any} X
     */
    PCA.prototype.fit = function (X) {
        ops_1.validateMatrix2D(X);
        ops_1.validateMatrixType(X, ['number']);
        var nSamples = X.length;
        // Renaming X to A for readability
        var A = tf.tensor2d(X);
        // const transposed = tf.transpose(A, [1, 0]);
        var AT = tf.transpose(A, [1, 0]);
        var M = tf.mean(AT, 1);
        var rawC = tf.sub(A, M);
        var C = ops_1.reshape(__spread(rawC.dataSync()), rawC.shape);
        var svd = numeric.svd(C);
        this.components = svd.V;
        this.explained_variance = numeric.div(numeric.pow(svd.U), nSamples - 1);
    };
    /**
     * Predict does nothing in PCA
     * @param X - A 2D matrix
     */
    PCA.prototype.predict = function (X) {
        if (X === void 0) { X = null; }
        console.info('Predict does nothing in PCA\n', X);
        return null;
    };
    /**
     * Saves the model's states
     */
    PCA.prototype.toJSON = function () {
        return {
            components: this.components,
            explained_variance: this.explained_variance
        };
    };
    /**
     * Restores the model from given states
     * @param components - Principal axes in feature space, representing the directions of maximum variance in the data.
     * @param explained_variance - The amount of variance explained by each of the selected components.
     */
    PCA.prototype.fromJSON = function (_a) {
        var _b = _a === void 0 ? {
            components: null,
            explained_variance: null
        } : _a, _c = _b.components, components = _c === void 0 ? null : _c, _d = _b.explained_variance, explained_variance = _d === void 0 ? null : _d;
        this.components = components;
        this.explained_variance = explained_variance;
    };
    return PCA;
}());
exports.PCA = PCA;
